### PR TITLE
v4 events and v3 support

### PR DESCRIPTION
**Pull request by TroyKomodo**
## Proposed changes

- v4 events
- convert back into v3 events in the event api

## Types of changes

What types of changes does your code introduce to 7TV?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/SevenTV/SevenTV/blob/main/CONTRIBUTING.md) doc
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

### CODE REVIEW

File: apps/api/src/http/v3/gql/mutations/reports.rs [hunk 1]
```diff
@@ -274,6 +204,45 @@ impl ReportsMutation {
...
+				tx.register_event(EventPayload {
+					actor_id: Some(auth_sesion.user_id()),
+					data: EventPayloadData::Ticket {
+						after: ticket.clone(),
+						data: event_ticket_data,
+					},
+					timestamp: chrono::Utc::now(),
+				})?;
+			}
+
```
Comment (TroyKomodo) [inline apps/api/src/http/v3/gql/mutations/reports.rs:245]: Is this truly what we want to do?

What if we only emit a single event per / mutation?


And we combine these multiple events into a single?

Comment (lennartkloock) [reply apps/api/src/http/v3/gql/mutations/reports.rs:245]: I tried to keep it as simple as possible. What would be the advantage of grouping multiple events into one?

File: shared/src/event.rs [hunk 2]
```diff
@@ -0,0 +1,791 @@
...
+			StoredEventUserData, StoredEventUserEditorData, StoredEventUserProfilePictureData, StoredEventUserSessionData,
+		},
+		ticket::{Ticket, TicketMessage, TicketPriority},
+		user::{
+			ban::UserBan, connection::UserConnection, editor::{UserEditor, UserEditorPermissions}, profile_picture::UserProfilePicture, session::UserSession, FullUser, User
+		},
+		Id,
+	},
+	event_api::{self, types::{ChangeField, ChangeFieldType}}, old_types::{ActiveEmoteModel, EmoteFlagsModel, EmotePartialModel, EmoteSetModel, UserConnectionModel, UserConnectionPartialModel, UserPartialModel},
+};
```
Comment (TroyKomodo) [inline shared/src/event.rs:28]: i hate this.

File: apps/mongo-typesense/src/typesense/handlers.rs [hunk 3]
```diff
@@ -179,7 +183,7 @@ default_impl!(
 );
 default_impl!(user_ban_template_batcher, typesense::UserBanTemplate, mongo::UserBanTemplate);
 default_impl!(user_ban_batcher, typesense::UserBan, mongo::UserBan);
-default_impl!(audit_log_batcher, typesense::AuditLog, mongo::AuditLog);
+default_impl!(audit_log_batcher, typesense::Event, mongo::StoredEvent);
```
Comment (TroyKomodo) [inline apps/mongo-typesense/src/typesense/handlers.rs:186]: rename this

**Pull request merged**